addons/binding/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/handler/PHCBridgeHandler.java

```diff
@@ -0,0 +1,710 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.phc.handler;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.TooManyListenersException;
+import java.util.concurrent.ConcurrentNavigableMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+
+import org.apache.commons.lang.ArrayUtils;
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.StopMoveType;
+import org.eclipse.smarthome.core.library.types.UpDownType;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseBridgeHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.State;
+import org.openhab.binding.phc.PHCBindingConstants;
+import org.openhab.binding.phc.internal.PHCHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import gnu.io.CommPortIdentifier;
+import gnu.io.PortInUseException;
+import gnu.io.RXTXPort;
+import gnu.io.SerialPort;
+import gnu.io.SerialPortEvent;
+import gnu.io.SerialPortEventListener;
+import gnu.io.UnsupportedCommOperationException;
+
+/**
+ * The {@link PHCBridgeHandler} is responsible for handling the serial Communication to and from the PHC Modules.
+ *
+ * @author Jonas Hohaus - Initial contribution
+ */
+
+public class PHCBridgeHandler extends BaseBridgeHandler implements SerialPortEventListener {
+
+    private Logger logger = LoggerFactory.getLogger(PHCHandler.class);
+
+    private static final int BAUD = 19200;
+    private static final int SEND_RETRY_COUNT = 15; // max count to send the same message
+    private static final int SEND_RETRY_TIME = 80; // time to wait for an acknowledge before send the message again
+
+    private InputStream serialIn = null;
+    private OutputStream serialOut = null;
+    short lastReceivedCrc;
+    byte[] messageFragment = null;
+    RXTXPort serialPort = null;
+
+    private Map<String, Boolean> toggleMap = new HashMap<String, Boolean>();
+    private ConcurrentSkipListMap<Long, QueueObject> queue = new ConcurrentSkipListMap<Long, QueueObject>();
```

Comment:
The method subMap (line 278) needs the `ConcurrentSkipListMap`. Maybe you have a better solution.